File: 938_rangeSumBST.py
# 938. Range Sum of BST
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def sumRange(root, low, high, sum_):
    if not root:
        return sum_
    
    if root.val>=low and root.val<=high:
        sum_+=root.val
    
    sum_ = sumRange(root.left, low, high, sum_)
    sum_ = sumRange(root.right, low, high, sum_)
    
    return sum_
    

class Solution(object):
    def rangeSumBST(self, root, low, high):
        """
        :type root: TreeNode
        :type low: int
        :type high: int
        :rtype: int
        """
        
        return sumRange(root, low, high, 0)

File: 938_rangeSumBST.cpp
// 938. Range Sum of BST

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 
 */

class Solution {
public:
    int rangeSum(TreeNode* &root, int low, int high, int &sum){
        
        if(root==NULL){
            return sum;
        }

        if(root->val>=low && root->val<=high){
            sum+=root->val;
        }

        sum = rangeSum(root->left, low, high,sum);
        sum = rangeSum(root->right, low, high,sum);

        return sum;
    }
    
    int rangeSumBST(TreeNode* root, int low, int high) {
        int sum = 0;
        // cout<<(*root).left<<" "<<root->left;
        sum = rangeSum(root, low, high, sum);
        return sum;
    }
};
